packages/palette/src/elements/FilterSelect/Components/FilterSelectContext.tsx

```diff
@@ -0,0 +1,156 @@
+import reject from "lodash.reject"
+import React, {
+  createContext,
+  useContext,
+  useLayoutEffect,
+  useReducer,
+} from "react"
+import { INITIAL_ITEMS_TO_SHOW } from "../../ShowMore"
+
+export interface Item {
+  label: string
+  value: string
+  [key: string]: string | number | boolean
+}
+
+interface FilterSelectContextProps {
+  items: Item[]
+  filteredItems: Item[]
+  initialItemsToShow: number
+  isFiltered: boolean
+  onChange: (state: {
+    items: FilterSelectContextProps["items"]
+    filteredItems: FilterSelectContextProps["filteredItems"]
+    selectedItems: FilterSelectContextProps["selectedItems"]
+    query: FilterSelectContextProps["query"]
+  }) => void
+  order: [[string, string], ["asc" | "desc", "asc" | "desc"]] // See: https://lodash.com/docs/4.17.15#orderBy
+  placeholder: string
+  query: string
+  renderItemLabel?: (item: any) => string
+  selectedItems: Item[]
+  setSelectedItems: (item: Item) => void
+  setQuery: (query: string) => void
+}
+
+export type FilterSelectState = Pick<
+  FilterSelectContextProps,
+  | "filteredItems"
+  | "initialItemsToShow"
+  | "items"
+  | "isFiltered"
+  | "onChange"
+  | "order"
+  | "placeholder"
+  | "renderItemLabel"
+  | "query"
+  | "selectedItems"
+>
+
+type Action =
+  | { type: "SET_QUERY"; payload: { query: string } }
+  | { type: "SET_SELECTED_ITEMS"; payload: { item: Item } }
+
+const filterSelectReducer = (state: FilterSelectState, action: Action) => {
+  switch (action.type) {
+    case "SET_QUERY": {
+      const { query } = action.payload
+
+      if (query === "") {
+        return {
+          ...state,
+          query,
+          isFiltered: false,
+          filteredItems: [],
+        }
+      }
+
+      const filteredItems = state.items.filter(({ label: name }) => {
+        return name.toLowerCase().includes(query.toLowerCase())
+      })
+
+      return {
+        ...state,
+        query,
+        isFiltered: true,
+        filteredItems,
+      }
+    }
+
+    case "SET_SELECTED_ITEMS": {
+      const isFound = !!state.selectedItems.find(
+        (item) => item.value === action.payload.item.value
+      )
+
+      const selectedItems = isFound
+        ? reject(state.selectedItems, { value: action.payload.item.value })
+        : [...state.selectedItems, action.payload.item]
+
+      return {
+        ...state,
+        selectedItems,
+      }
+    }
+  }
+}
+
+const initialState: FilterSelectState = {
+  filteredItems: [],
+  initialItemsToShow: INITIAL_ITEMS_TO_SHOW,
+  isFiltered: false,
+  items: [],
+  onChange: (x) => x,
+  order: [
+    ["age", "name"],
```

Comment:
Seems odd this would have opinions on fields to sort by?

Reply:
ahh sorry, i copied this over from the example as i was tidying up. Lemme update